docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+- we should define also how credential for the InfraLoadBalancer should be provided
+- we should brainstorm both on the clusterctl workflow and on the github workflow (without clusterctl)
+- the “quick start” workflow should be as simple as possible  
+
+### Security Model
+
+Roles
+For the purposes of this security model, 3 common roles have been identified:
+Infrastructure provider: The infrastructure provider (infra) is responsible for the overall environment that the cluster(s) are operating in or the PaaS provider in a company.
+Management cluster operator: The cluster operator (ops) is responsible for administration of the Cluster API management cluster. They manage policies, network access, application permissions.
+Workload cluster operator: The workload cluster operator (dev) is responsible for management of the cluster relevant to their particular applications .
+
+#### Write Permissions
+
+| | Cluster | Load Balancer | AWSELBLoadBalancer | AWS IAM API | Relevant AWS APIs |
+|---|---|---|---|---|---|
+| Load Balancer Provider | No | No | Yes | No | Yes |
+| Load Balancer Controller | No | Yes | Yes | No | No |
+| Cluster Controller | Yes | Yes | No | No | No |
+| Management Cluster Operators | Yes | Yes | Yes | Yes | No |
+| Workload Cluster Operator | Yes | Yes | Yes | No | No |
+
+TODO: Credential management changes for infra providers
+
+#### Credential Management
+
+Infra load balancers need a specific credential workflow:
+Look at the infra load balancer resource for spec.CredentialRef:
+If it’s set: use it for load balancing provisioning
+If it’s not set: the infra load balancer provider tries to read credentials from the infrastructure cluster owning it:
+If the infra cluster is from the same type, use its credentialRef
+If the infra cluster is not from the same type, fallback to the credentials of infrastructure load balancer controller
+
+### Risks and Mitigations
+
+- What are the risks of this proposal and how do we mitigate? Think broadly.
+- How will UX be reviewed and by whom?
+- How will security be reviewed and by whom?
+- Consider including folks that also work outside the SIG or subproject.
+
+## Alternatives
+
+### Using Services API
+
+Kubernetes already provides a load balancer construct of Service Type=Loadbalancer. The service API allows pointing at arbitrary IPs, which could in fact be the machines of a workload cluster’s control plane.
+
+While we believe this approach has longer term benefits, there are a few reasons that it does not fit the short term requirements:
+
+- Existing Service Type=LoadBalancer implementations do not support credential multi-tenancy
+
+- Existing Service Type=LoadBalancer implementations do not support co-deployment with other implementations
+
+- ??? (I know there were other reasons, but blanking on them at the moment, need to finish out this list)
```

Comment:
Can you use an AWS CCM to implement SotLB if the cluster itself is running on vSphere/VMC? Does that need to go here? (Or rather a generic version of)